### DISCUSSION TITLE
Make the node_modules folder more space efficient by having a --production flag to install only the files used

### DISCUSSION THREAD

**Discussion by destroyer22719** [npm/rfcs#292] [RRFC] [answered]
Title sums it all. We should have a `npm install <package>` with an `npm -P install <package>` which would only install the files from say `@angular/core` that I used in my code. Not the entire package. This can help reduce the size of the node_modules folder.

**Comment by ljharb** [comment]
How would npm determine that, exactly?

**Comment by darcyclarke** [comment, marked as answer]
@destroyer22719 Appreciate the suggestion. After some discussion in a recent RFC call, I don't think there's an easy way to support this today. Unfortunately, we don't tend to know the intent of the package author; ie. potentially an author wants to ship files that aren't used at runtime (ex. licenses, static files/assets to be copied etc.) so it's hard to determine what is or isn't used or what should or should not be installed. Static code analysis of your project would also only tell you so much about usage.

That said, one way to solve this issue today is, ask package authors to only ship the necessary files required for usage of that package (which may mean they end up publishing different packages for those different use cases). Another concept we have floated in this realm would be to intercept module requests and introduce a virtual file system (ex. `tink`/`yarn` **pnp**) and this actually may be something we look at supporting in the future (this would also mean no `install` is required). Lastly, we're considering supporting a reify strategy that is similar to `pnpm`, where you would only ever install one copy of a module (helping with overall disk space); That RFC is currently being worked on but should be ready early into the new year.

Ultimately, I don't think we'd want to change the current behavior of `npm install` in the way you've suggested but are expecting there will be more space-saving options for you in the near future.

**Reply by destroyer22719** [reply]
Hi! Thank you so much for your response, my idea is `npm install --production` would only install filles that are imported, and what the imported files imported. Maybe this will exclude unused `.js` or `.ts` files. That are defined in your `package.json` file. At the end of the day I understand your reasoning and thank you so much for listening to my suggestion.
